tools/autograd/gen_variable_type.py

```diff
@@ -771,19 +768,26 @@ def emit_body(declaration):
 
         env['check_inplace'] = ''
         env['version_counter'] = ''
+        maybe_inplace = any(arg['name'] == 'inplace' for arg in arguments)
         base_call = BASE_CALL.substitute(combined)
         if declaration['inplace']:
             env['check_inplace'] = 'check_inplace(self);'
             env['version_counter'] = 'increment_version(self);'
-        elif any(arg['name'] == 'inplace' for arg in arguments):
+            modifies_data = 'false' if is_view else 'true'
+            env['set_flags'] = SET_FLAGS_INPLACE.substitute(combined, modifies_data=modifies_data)
+        elif maybe_inplace:
             assert not is_view, declaration['name']
             env['check_inplace'] = 'if (inplace) check_inplace(input);'
             env['version_counter'] = 'if (inplace) increment_version(input);'
+            env['set_flags'] = SET_FLAGS_INPLACE.substitute(combined, modifies_data='inplace')
+            base_call = 'auto ret = maybe_wrap({}, input, inplace)'.format(base_call)
+        else:
+            env['set_flags'] = SET_FLAGS.substitute(combined)
+            if is_view:
+                base_call = 'auto ret = as_view(static_cast<const Variable&>(self), {})'.format(base_call)
+            else:
+                base_call = 'auto ret = as_variable({})'.format(base_call)
```

Comment:
Is there any significant change in the logic here, or have you just rearranged things? Is it the addition of `maybe_wrap` (and why is it necessary)?